README.md
# NAME

Plack::I18N - I18N for Plack

# SYNOPSIS

    use Plack::I18N;
    use Plack::Builder;

    my $i18n = Plack::I18N->new(lexicon => 'gettext', locale_dir => 'locale/');

    builder {
        enable 'I18N', i18n => $i18n;

        sub {
            my $env = shift;

            my $handle = $env->{'plack.i18n.handle'};

            [200, [], [$handle->maketext('Hello')]];
        };
    };

# DESCRIPTION

Plack::I18N is an easy way to add i18n to your application. Plack::I18N supports
both [Locale::Maketext](https://metacpan.org/pod/Locale::Maketext) `*.pm` files and `gettext` `*.po` files. Use
whatevers suits better.

See [https://github.com/vti/plack-i18n/tree/master/examples](https://github.com/vti/plack-i18n/tree/master/examples) directory for both
examples.

## Language detection

Language detection is done via HTTP headers, session cookies, URL path prefix
and so on. See [Plack::Middleware::I18N](https://metacpan.org/pod/Plack::Middleware::I18N) for details.

## `$env` parameters

Plack::Middleware::I18N registers the following `$env` parameters:

- `plack.i18n`

    Holds Plack::I18N instance.

- `plack.i18n.language`

    Current detected language. A shortcut for `$env-`{'plack.i18n'}->language>.

- `plack.i18n.handle`

    A shortcut for `$env-`{'plack.i18n'}->handle($env->{'plack.i18n.language'})>.

# METHODS

## `new`

Creates new object.

## `default_language`

Returns default language.

## `handle($language)`

Returns handle of appropriate language.

## `languages`

Returns available languages.

# AUTHOR

Viacheslav Tykhanovskyi, <viacheslav.t@gmail.com>

# COPYRIGHT AND LICENSE

Copyright (C) 2015, Viacheslav Tykhanovskyi

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

This program is distributed in the hope that it will be useful, but without any
warranty; without even the implied warranty of merchantability or fitness for
a particular purpose.


Changes
Revision history for Perl extension Plack-I18N

{{$NEXT}}

0.04 2015-02-02T08:26:57Z

    - fix order in tests
    - more docs

0.03 2015-02-02T07:26:22Z

    - use abs paths in tests

0.02 2015-02-01T11:32:15Z

    - fix dist meta information
    - more reference documentation

0.01 2015-01-30T15:04:12Z

    - original version



lib/Plack/I18N/Lexicon/Gettext.pm
package Plack::I18N::Lexicon::Gettext;

use strict;
use warnings;

use Carp qw(croak);
use List::Util qw(first);
use Locale::Maketext;
use Plack::I18N::Util qw(try_load_class);
use Plack::I18N::Handle;

sub new {
    my $class = shift;
    my (%params) = @_;

    my $self = {};
    bless $self, $class;

    $self->{i18n_class} = $params{i18n_class} || croak 'i18n_class required';
    $self->{default_language} = $params{default_language} || 'en';

    $self->{locale_dir} = $params{locale_dir} || croak 'locale_dir required';

    $self->_init_lexicon;

    return $self;
}

sub _init_lexicon {
    my $self = shift;

    my $i18n_class = $self->{i18n_class};

    if (!try_load_class($i18n_class)) {
        eval <<"EOC" || croak $@;
        package $i18n_class;
        use parent 'Locale::Maketext';
        use Locale::Maketext::Lexicon {
            '*'      => [Gettext => "$self->{locale_dir}/*.po"],
            _auto    => 1,
            _decode  => 1,
            _preload => 1
        };
        1;
EOC
    }
}

sub detect_languages {
    my $self = shift;

    my $path = $self->{locale_dir};

    opendir(my $dh, $path) or croak "Can't opendir $path: $!";
    my @files = grep { /\.p[om]$/ && -e "$path/$_" } readdir($dh);
    closedir $dh;

    my @languages = @files;
    s{\.p[om]$}{} for @languages;

    unshift @languages, $self->{default_language}
      unless first { $_ eq $self->{default_language} } @languages;

    return @languages;
}

1;
__END__
=pod

=encoding utf-8

=head1 NAME

Plack::I18N::Lexicon::Gettext - Module

=head1 SYNOPSIS



=head1 DESCRIPTION



=head1 METHODS

=head2 C<new>

=head2 C<detect_languages>

=head1 AUTHOR

Viacheslav Tykhanovskyi, E<lt>viacheslav.t@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015, Viacheslav Tykhanovskyi

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

This program is distributed in the hope that it will be useful, but without any
warranty; without even the implied warranty of merchantability or fitness for
a particular purpose.

=cut


lib/Plack/I18N/Lexicon/Maketext.pm
package Plack::I18N::Lexicon::Maketext;

use strict;
use warnings;

use Carp qw(croak);
use List::Util qw(first);
use Locale::Maketext;
use Plack::I18N::Util qw(try_load_class);
use Plack::I18N::Handle;

sub new {
    my $class = shift;
    my (%params) = @_;

    my $self = {};
    bless $self, $class;

    $self->{i18n_class} = $params{i18n_class} || croak 'i18n_class required';
    $self->{locale_dir} = $params{locale_dir};
    $self->{default_language} = $params{default_language} || 'en';

    $self->_init_lexicon;

    return $self;
}

sub _init_lexicon {
    my $self = shift;

    my $i18n_class = $self->{i18n_class};

    if (!try_load_class($i18n_class)) {
        eval <<"EOC" or croak $@;
            package $i18n_class;
            use parent 'Locale::Maketext';
            sub _loaded {1}
            1;
EOC
    }

    my $default_i18n_class = "$i18n_class\::$self->{default_language}";
    if (!try_load_class($default_i18n_class)) {
        eval <<"EOC" or croak $@;
            package $default_i18n_class;
            use parent -norequire, '$i18n_class';
            our %Lexicon = (_AUTO => 1);
            sub _loaded {1}
            1;
EOC
    }
}

sub detect_languages {
    my $self = shift;

    my $path = $self->{locale_dir} || $INC{$self->{i18n_class}} || join '/',
      'lib', split /::/, $self->{i18n_class};

    opendir(my $dh, $path) or croak "Can't opendir $path: $!";
    my @files = grep { /\.p[om]$/ && -e "$path/$_" } readdir($dh);
    closedir $dh;

    my @languages = @files;
    s{\.pm$}{} for @languages;

    unshift @languages, $self->{default_language}
      unless first { $_ eq $self->{default_language} } @languages;

    return @languages;
}

1;
__END__
=pod

=encoding utf-8

=head1 NAME

Plack::I18N::Lexicon::Maketext - Module

=head1 SYNOPSIS



=head1 DESCRIPTION



=head1 METHODS

=head2 C<new>

=head2 C<detect_languages>

=head1 AUTHOR

Viacheslav Tykhanovskyi, E<lt>viacheslav.t@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015, Viacheslav Tykhanovskyi

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

This program is distributed in the hope that it will be useful, but without any
warranty; without even the implied warranty of merchantability or fitness for
a particular purpose.

=cut


lib/Plack/I18N.pm
package Plack::I18N;
use 5.008001;
use strict;
use warnings;

our $VERSION = "0.03";

use Carp qw(croak);
use Locale::Maketext;
use Plack::Util ();
use Plack::I18N::Handle;

sub new {
    my $class = shift;
    my (%params) = @_;

    croak 'lexicon required'    unless $params{lexicon};

    my $self = {};
    bless $self, $class;

    $self->{i18n_class}       = $params{i18n_class} || __PACKAGE__ . '::Auto';
    $self->{locale_dir}       = $params{locale_dir};
    $self->{default_language} = $params{default_language} || 'en';

    $self->{lexicon} = $self->_build_lexicon(
        $params{lexicon},
        i18n_class       => $self->{i18n_class},
        locale_dir       => $self->{locale_dir},
        default_language => $self->{default_language}
    );

    $self->{languages} =
      $params{languages} || [sort $self->{lexicon}->detect_languages];

    return $self;
}

sub default_language {
    my $self = shift;

    return $self->{default_language};
}

sub languages {
    my $self = shift;

    return @{$self->{languages}};
}

sub handle {
    my $self = shift;
    my ($language) = @_;

    my $i18n_class = $self->{i18n_class};

    $self->{handles}->{$language} ||= do {
        my $handle = $i18n_class->get_handle($language);
        $handle->fail_with(sub { $_[1] });

        Plack::I18N::Handle->new(handle => $handle, language => $language);
    };

    return $self->{handles}->{$language};
}

sub _build_lexicon {
    my $self = shift;
    my ($lexicon, %params) = @_;

    my $lexicon_class = __PACKAGE__ . '::Lexicon::' . ucfirst($lexicon);

    Plack::Util::load_class($lexicon_class);

    return $lexicon_class->new(%params);
}

1;
__END__
=pod

=encoding utf-8

=head1 NAME

Plack::I18N - I18N for Plack

=head1 SYNOPSIS

    use Plack::I18N;
    use Plack::Builder;

    my $i18n = Plack::I18N->new(lexicon => 'gettext', locale_dir => 'locale/');

    builder {
        enable 'I18N', i18n => $i18n;

        sub {
            my $env = shift;

            my $handle = $env->{'plack.i18n.handle'};

            [200, [], [$handle->maketext('Hello')]];
        };
    };

=head1 DESCRIPTION

Plack::I18N is an easy way to add i18n to your application. Plack::I18N supports
both L<Locale::Maketext> C<*.pm> files and C<gettext> C<*.po> files. Use
whatevers suits better.

See L<https://github.com/vti/plack-i18n/tree/master/examples> directory for both
examples.

=head2 Language detection

Language detection is done via HTTP headers, session cookies, URL path prefix
and so on. See L<Plack::Middleware::I18N> for details.

=head2 C<$env> parameters

Plack::Middleware::I18N registers the following C<$env> parameters:

=over

=item C<plack.i18n>

Holds Plack::I18N instance.

=item C<plack.i18n.language>

Current detected language. A shortcut for C<$env-E<gt>{'plack.i18n'}-E<gt>language>.

=item C<plack.i18n.handle>

A shortcut for C<$env-E<gt>{'plack.i18n'}-E<gt>handle($env-E<gt>{'plack.i18n.language'})>.

=back

=head1 METHODS

=head2 C<new>

Creates new object.

Options:

=over

=item lexicon

One of C<gettext> or C<maketext>.

=item i18n_class

This is usually C<MyApp::I18N>. This class is automatically generated if
does not exist. In case of C<gettext> C<i18n_class> it even doesn't
have to be specified.

=item locale_dir

Directory where translations are stored.

=item default_language

Default language. C<en> by default.

=item languages

Available languages. Automatically detected unless specified.

=back

=head2 C<default_language>

Returns default language.

=head2 C<handle($language)>

Returns handle of appropriate language.

=head2 C<languages>

Returns available languages.

=head1 AUTHOR

Viacheslav Tykhanovskyi, E<lt>viacheslav.t@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015, Viacheslav Tykhanovskyi

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

This program is distributed in the hope that it will be useful, but without any
warranty; without even the implied warranty of merchantability or fitness for
a particular purpose.

=cut
